# Combining DataFrames With Pandas - Lab

## Introduction

In this lab, we'll gain practice combining DataFrames through concatenation.  We'll also learn to read in tables from a database and store them in separate tables, as well as how to execute various types of joins to selectively combine the information stored in the tables!


## Objectives:
You will be able to:
* Understand and explain when to use DataFrame joins and merges
* Be able to use pd.merge when combining DataFrames based on column values
* Understand, explain and use a range of DataFrame merge types: outer, inner, left and right
* Use pd.concat() to stack DataFrames

## Getting Started

We'll start with a quick section to help us gain practice with concatenating datasets using `pd.concat()`.

### Concatenating DataFrames

Run the cell below to create some sample DataFrames for us to work with.  

In [1]:
import pandas as pd
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                    index=[0, 1, 2, 3])


df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                    index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'], 
                    'D': ['D8', 'D9', 'D10', 'D11']},
                    index=[8, 9, 10, 11])

Now that we have multiple DataFrames to work with, we can execute a concatenation to join them together.  

In the cell below, concatenate the 3 DataFrames together using the appropriate function.  

In [2]:
combined_df = pd.concat([df1, df2, df3])
combined_df

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

**_EXPECTED OUTPUT:_**

<img src="er1.png">

### Setting Join Conditions With Concatenation

We can also specify if the concatenation is an **_Outer Join_** or an **_Inner Join_**.  Next, we'll execute an inner join. Before we do, we need to create another table that contains some overlapping index values with a DataFrame that already exists.  

Run the cell below to create the new DataFrame.

In [3]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'],
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                    index=[2, 3, 6, 7])

In [4]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [5]:
df4

B   D   F
2  B2  D2  F2
3  B3  D3  F3
6  B6  D6  F6
7  B7  D7  F7

Now, in the cell below, use the `pd.concat()` method to join DataFrames 1 and 4.  However, this time, specify that the `join` is `'inner'`, and `axis=1`. 

In [6]:
df1_and_4 = None
df1_and_4 = pd.concat([df1, df4], join='inner', axis=1)

df1_and_4

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

**_EXPECTED OUTPUT:_**

<img src='er2.png'>

We'll notice that in this case, the results returned contain only the rows with indexes that exist in both tables--rows 2 and 3.  The resulting table contains the values for each column in both tables for the rows.  

Note that there are many, many ways that we can make full use of the `concat()` functionality in pandas to join DataFrames together--these are just a few of the most common examples pulled from the pandas documentation on the subject.  For a full view of all the ways we can use `pd.concat()`, see the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html)!

## Loading In Data From SQL Database Tables

Now, we'll move on to working with the Hearthstone cards database.  This database contains information on cards from the popular game, [Hearthstone](https://playhearthstone.com/en-us/)! For full information on the dataset, see the  [kaggle page](https://www.kaggle.com/jeradrose/hearthstone-cards) for this dataset. 

This database consists of the following tables:

* _cards_
* *dust_costs*
* _entourages_
* _mechanics_
* *play_requirements*

Many of rows in each table--but not all--correspond to the same cards. As such, each table contains a column called `card_id` which acts as a **_Primary Key_** for each table.  We'll make use of these keys to **_join_** the different tables together into a single DataFrame.  We'll also experiment with different types of joins to help us decide exactly what information we want to combine.  

Before we can begin working with the database, we'll first need to connect to it.  For this, we'll use the `sqlalchemy` library.

In the cell below:

* Import the `create_engine` module from `sqlalchemy`
* Create an engine that connects to `sqlite:///database.sqlite`.

In [7]:
from sqlalchemy import create_engine

# creates engine object
engine = create_engine('sqlite:///database.sqlite')

In [8]:
engine

Engine(sqlite:///database.sqlite)

Now that we've successfully connected to our Database, we'll make use of pandas to read in each table individually.  

In the cell below, read in each sql table into separate pandas DataFrames using the `read_sql_table` method.  

The first argument should be a string corresponding to the name of the table, and the second argument should be the `engine` object we created in the cell above. 

If you're unsure of how to do this, see the [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_sql_table.html).

In [9]:
# Lee - ???
cards_df = pd.read_sql_table('cards', engine)

In [10]:
len(cards_df)

2819

In [11]:
cards_df

card_id player_class         type  \
0                        KARA_00_07         MAGE        SPELL   
1                         NEW1_008a        DRUID        SPELL   
2                         BRM_010t2        DRUID       MINION   
3                            AT_132      NEUTRAL       MINION   
4                            OG_141      NEUTRAL       MINION   
5                           CRED_17      NEUTRAL       MINION   
6                           OG_337e      NEUTRAL  ENCHANTMENT   
7                         BRMA08_1H      NEUTRAL         HERO   
8                          GVG_011a      NEUTRAL  ENCHANTMENT   
9                       NAX8_02H_TB      NEUTRAL   HERO_POWER   
10                           AT_131      NEUTRAL       MINION   
11                         CS2_122e      NEUTRAL  ENCHANTMENT   
12                      LOEA05_02ha      NEUTRAL   HERO_POWER   
13                CS2_102_H1_AT_132      WARRIOR   HERO_POWER   
14                         CFM_334e      NEUTRAL  ENCHANTMENT   
15                          EX1_583      NEUTRAL       MINION   
16           TB_PickYourFate_11rand      NEUTRAL        SPELL   
17                         LOE_007t      WARLOCK        SPELL   
18                          CFM_095      NEUTRAL       MINION   
19                          BRM_018      PALADIN       MINION   
20                         PRO_001a      NEUTRAL        SPELL   
21                        BRMA10_5H      NEUTRAL       MINION   
22                         BRMA11_1      NEUTRAL         HERO   
23                           OG_027       SHAMAN        SPELL   
24                          NAX1_01      NEUTRAL         HERO   
25                          GVG_030        DRUID       MINION   
26                          CFM_094      WARLOCK        SPELL   
27                      DS1h_292_H1       HUNTER   HERO_POWER   
28                          CS2_227      NEUTRAL       MINION   
29                          KAR_062      NEUTRAL       MINION   
...                             ...          ...          ...   
2789                       LOEA12_2      NEUTRAL   HERO_POWER   
2790                         AT_084      NEUTRAL       MINION   
2791                     KARA_00_03         MAGE         HERO   
2792                       BRMA16_3      NEUTRAL        SPELL   
2793                        tt_010a         MAGE       MINION   
2794                        EX1_559         MAGE       MINION   
2795                        EX1_624       PRIEST        SPELL   
2796                        OG_044a        DRUID       MINION   
2797                         AT_044        DRUID        SPELL   
2798                  AT_132_SHAMAN       SHAMAN   HERO_POWER   
2799                     CFM_621t29      NEUTRAL        SPELL   
2800                       EX1_355e      PALADIN  ENCHANTMENT   
2801                      LOEA09_3d      NEUTRAL   HERO_POWER   
2802                  TB_Coopv3_105      NEUTRAL       MINION   
2803                       BRMA13_4      NEUTRAL   HERO_POWER   
2804                       EX1_345t       PRIEST       MINION   
2805                        EX1_009      NEUTRAL       MINION   
2806                        CRED_32      NEUTRAL       MINION   
2807                       FP1_030e      NEUTRAL  ENCHANTMENT   
2808        TB_PickYourFate_10_Ench      NEUTRAL  ENCHANTMENT   
2809             KARA_05_01hpheroic      NEUTRAL   HERO_POWER   
2810                       BRMA10_3      NEUTRAL   HERO_POWER   
2811  TB_PickYourFate_10_EnchMinion      NEUTRAL  ENCHANTMENT   
2812                       NAX3_02H      NEUTRAL   HERO_POWER   
2813                        AT_083e      NEUTRAL  ENCHANTMENT   
2814                        AT_066e      WARRIOR  ENCHANTMENT   
2815                        CS2_031         MAGE        SPELL   
2816                        OG_195e      NEUTRAL  ENCHANTMENT   
2817                       XXX_111e      NEUTRAL  ENCHANTMENT   
2818                        NAX3_01      NEUTRAL         HERO  

In [12]:
dust_df = pd.read_sql_table('dust_costs', engine)

In [13]:
len(dust_df)

4040

In [14]:
dust_df

card_id             action  cost
0     BRM_010t2    CRAFTING_NORMAL    40
1     BRM_010t2    CRAFTING_GOLDEN   400
2     BRM_010t2  DISENCHANT_NORMAL     5
3     BRM_010t2  DISENCHANT_GOLDEN    50
4        AT_132    CRAFTING_NORMAL  1600
5        AT_132    CRAFTING_GOLDEN  3200
6        AT_132  DISENCHANT_NORMAL   400
7        AT_132  DISENCHANT_GOLDEN  1600
8        OG_141    CRAFTING_NORMAL    40
9        OG_141    CRAFTING_GOLDEN   400
10       OG_141  DISENCHANT_NORMAL     5
11       OG_141  DISENCHANT_GOLDEN    50
12       AT_131    CRAFTING_NORMAL  1600
13       AT_131    CRAFTING_GOLDEN  3200
14       AT_131  DISENCHANT_NORMAL   400
15       AT_131  DISENCHANT_GOLDEN  1600
16      EX1_583    CRAFTING_NORMAL    40
17      EX1_583    CRAFTING_GOLDEN   400
18      EX1_583  DISENCHANT_NORMAL     5
19      EX1_583  DISENCHANT_GOLDEN    50
20      CFM_095    CRAFTING_NORMAL   400
21      CFM_095    CRAFTING_GOLDEN  1600
22      CFM_095  DISENCHANT_NORMAL   100
23      CFM_095  DISENCHANT_GOLDEN   400
24      BRM_018    CRAFTING_NORMAL   100
25      BRM_018    CRAFTING_GOLDEN   800
26      BRM_018  DISENCHANT_NORMAL    20
27      BRM_018  DISENCHANT_GOLDEN   100
28       OG_027    CRAFTING_NORMAL   100
29       OG_027    CRAFTING_GOLDEN   800
...         ...                ...   ...
4010    tt_010a  DISENCHANT_NORMAL   100
4011    tt_010a  DISENCHANT_GOLDEN   400
4012    EX1_559    CRAFTING_NORMAL  1600
4013    EX1_559    CRAFTING_GOLDEN  3200
4014    EX1_559  DISENCHANT_NORMAL   400
4015    EX1_559  DISENCHANT_GOLDEN  1600
4016    EX1_624    CRAFTING_NORMAL   100
4017    EX1_624    CRAFTING_GOLDEN   800
4018    EX1_624  DISENCHANT_NORMAL    20
4019    EX1_624  DISENCHANT_GOLDEN   100
4020    OG_044a    CRAFTING_NORMAL    40
4021    OG_044a    CRAFTING_GOLDEN   400
4022    OG_044a  DISENCHANT_NORMAL     5
4023    OG_044a  DISENCHANT_GOLDEN    50
4024     AT_044    CRAFTING_NORMAL   400
4025     AT_044    CRAFTING_GOLDEN  1600
4026     AT_044  DISENCHANT_NORMAL   100
4027     AT_044  DISENCHANT_GOLDEN   400
4028   EX1_345t    CRAFTING_NORMAL   400
4029   EX1_345t    CRAFTING_GOLDEN  1600
4030   EX1_345t  DISENCHANT_NORMAL   100
4031   EX1_345t  DISENCHANT_GOLDEN   400
4032    EX1_009    CRAFTING_NORMAL   100
4033    EX1_009    CRAFTING_GOLDEN   800
4034    EX1_009  DISENCHANT_NORMAL    20
4035    EX1_009  DISENCHANT_GOLDEN   100
4036    CS2_031    CRAFTING_NORMAL    40
4037    CS2_031    CRAFTING_GOLDEN   400
4038    CS2_031  DISENCHANT_NORMAL     5
4039    CS2_031  DISENCHANT_GOLDEN    50

[4040 rows x 3 columns]

In [16]:
entourages_df = pd.read_sql_table('entourages', engine)

In [17]:
len(entourages_df)

328

In [18]:
entourages_df

card_id       entourage_card_id
0                KAR_A10_22              KAR_A10_09
1                KAR_A10_22              KAR_A10_02
2                KAR_A10_22              KAR_A10_08
3                KAR_A10_22              KAR_A10_04
4                KAR_A10_22              KAR_A10_05
5                  TBST_006                TBST_003
6                  TBST_006                TBST_002
7                  TBST_006                TBST_001
8                BRMA14_10H                BRMA14_3
9                BRMA14_10H               BRMA14_5H
10               BRMA14_10H               BRMA14_7H
11               BRMA14_10H               BRMA14_9H
12                 BRMA09_3                BRMA09_2
13                 BRMA09_3                BRMA09_4
14                 BRMA09_3                BRMA09_5
15             KARA_06_03hp              KARA_06_01
16                 BRMA12_2                BRMA12_6
17                 BRMA12_2                BRMA12_5
18                 BRMA12_2                BRMA12_7
19                 BRMA12_2                BRMA12_4
20                 BRMA12_2                BRMA12_3
21      TB_PickYourFate_2nd       TB_PickYourFate_1
22      TB_PickYourFate_2nd       TB_PickYourFate_4
23      TB_PickYourFate_2nd   TB_PickYourFate_7_2nd
24      TB_PickYourFate_2nd       TB_PickYourFate_3
25      TB_PickYourFate_2nd   TB_PickYourFate_6_2nd
26      TB_PickYourFate_2nd  TB_PickYourFate_11rand
27                   TB_001               BRMA14_10
28                   TB_001              BRMA14_10H
29                   TB_001                BRMA09_4
..                      ...                     ...
298              LOEA16_16H               LOEA16_10
299              LOEA16_16H               LOEA16_11
300              LOEA16_16H               LOEA16_14
301              LOEA16_16H               LOEA16_15
302              LOEA16_16H                LOEA16_6
303              LOEA16_16H                LOEA16_7
304              LOEA16_16H                LOEA16_9
305              LOEA16_16H               LOEA16_12
306              LOEA16_16H               LOEA16_13
307              LOEA16_16H                LOEA16_8
308                  TBUD_1                 CS1_042
309                  TBUD_1                 CS2_171
310               BRMA13_3H                BRMA13_8
311               BRMA13_3H                BRMA13_7
312               BRMA13_3H                BRMA13_6
313               BRMA13_3H                BRMA13_5
314  TB_MechWar_CommonCards                 GVG_069
315  TB_MechWar_CommonCards                 GVG_079
316  TB_MechWar_CommonCards                 GVG_034
317  TB_MechWar_CommonCards                 GVG_048
318  TB_MechWar_CommonCards                 GVG_049
319  TB_MechWar_CommonCards                 GVG_088
320              KARA_00_10                 EX1_294
321              KARA_00_10                 FP1_018
322              KARA_00_10                  AT_002
323              KARA_00_10                 EX1_287
324              KARA_00_10                  tt_010
325              KARA_00_10                 EX1_295
326              KARA_00_10                 EX1_289
327              KARA_00_10                 EX1_594

[328 rows x 2 columns]

In [19]:
mechanics_df = pd.read_sql_table('mechanics', engine)

In [20]:
len(mechanics_df)

1079

In [21]:
mechanics_df

card_id             mechanic
0                 AT_132            BATTLECRY
1               GVG_011a  TAG_ONE_TURN_EFFECT
2                EX1_583            BATTLECRY
3               LOE_007t            EVIL_GLOW
4               LOE_007t   ImmuneToSpellpower
5                CFM_095          DEATHRATTLE
6                BRM_018            BATTLECRY
7                GVG_030           CHOOSE_ONE
8                GVG_030                TAUNT
9                CS2_227                 AURA
10               KAR_062            BATTLECRY
11           CFM_712_t17           JADE_GOLEM
12              BRM_011t  TAG_ONE_TURN_EFFECT
13             BRMA10_3H         AI_MUST_PLAY
14              KAR_710m                TAUNT
15              EX1_295o  TAG_ONE_TURN_EFFECT
16                OG_158          DEATHRATTLE
17              LOE_110t              TOPDECK
18              LOE_110t   ImmuneToSpellpower
19             BRMA13_4H         AI_MUST_PLAY
20              EX1_145o  TAG_ONE_TURN_EFFECT
21             LOEA09_2e  TAG_ONE_TURN_EFFECT
22              EX1_570e  TAG_ONE_TURN_EFFECT
23               GVG_069            BATTLECRY
24               hexfrog                TAUNT
25                OG_145        DIVINE_SHIELD
26                OG_145                TAUNT
27                AT_090              INSPIRE
28                AT_060               SECRET
29    TB_MechWar_Minion1                TAUNT
...                  ...                  ...
1049             EX1_284            BATTLECRY
1050             LOE_038                 AURA
1051              AT_083              INSPIRE
1052    TB_CoOpv3_BOSS2e              ENRAGED
1053             GVG_009            BATTLECRY
1054             CFM_338            BATTLECRY
1055             GVG_097            BATTLECRY
1056             GVG_097                TAUNT
1057             CFM_852            BATTLECRY
1058             LOE_020            BATTLECRY
1059             CFM_781              STEALTH
1060             GVG_090            BATTLECRY
1061             GVG_120            BATTLECRY
1062            EX1_055o  TAG_ONE_TURN_EFFECT
1063             CFM_321            BATTLECRY
1064             EX1_082            BATTLECRY
1065             FP1_017                 AURA
1066             FP1_001          DEATHRATTLE
1067              AT_089              INSPIRE
1068             CFM_646          DEATHRATTLE
1069              AT_084            BATTLECRY
1070             OG_044a               CHARGE
1071             OG_044a                TAUNT
1072           LOEA09_3d         AI_MUST_PLAY
1073            BRMA13_4         AI_MUST_PLAY
1074             EX1_009              ENRAGED
1075            BRMA10_3         AI_MUST_PLAY
1076            NAX3_02H         AI_MUST_PLAY
1077             AT_083e  TAG_ONE_TURN_EFFECT
1078             CS2_031               FREEZE

[1079 rows x 2 columns]

In [22]:
play_requirements_df = pd.read_sql_table('play_requirements', engine)

In [23]:
len(play_requirements_df)

776

In [24]:
play_requirements_df

card_id               play_requirement  value
0           KARA_00_07           REQ_NUM_MINION_SLOTS      1
1             PRO_001a           REQ_NUM_MINION_SLOTS      1
2              NAX1_01           REQ_NUM_MINION_SLOTS      1
3          DS1h_292_H1                REQ_STEADY_SHOT      0
4          DS1h_292_H1       REQ_MINION_OR_ENEMY_HERO      0
5              KAR_076             REQ_TARGET_TO_PLAY      0
6               AT_024             REQ_TARGET_TO_PLAY      0
7               AT_024           REQ_TARGET_WITH_RACE     15
8               AT_024              REQ_MINION_TARGET      0
9             BRMA06_2           REQ_NUM_MINION_SLOTS      1
10             NAX5_02      REQ_MINIMUM_ENEMY_MINIONS      1
11             OG_080c            REQ_FRIENDLY_TARGET      0
12             OG_080c             REQ_TARGET_TO_PLAY      0
13             OG_080c              REQ_MINION_TARGET      0
14             CS2_039             REQ_TARGET_TO_PLAY      0
15             CS2_039              REQ_MINION_TARGET      0
16   TB_KaraPortal_002              REQ_MINION_TARGET      0
17             CS2_094             REQ_TARGET_TO_PLAY      0
18     TB_SPT_DPromoHP              REQ_MINION_TARGET      0
19            BRMA15_3           REQ_NUM_MINION_SLOTS      1
20             CS2_141        REQ_TARGET_IF_AVAILABLE      0
21             CS2_141             REQ_NONSELF_TARGET      0
22             CS2_063             REQ_TARGET_TO_PLAY      0
23             CS2_063              REQ_MINION_TARGET      0
24             CS2_063               REQ_ENEMY_TARGET      0
25            GVG_055e            REQ_FRIENDLY_TARGET      0
26              TB_006             REQ_TARGET_TO_PLAY      0
27              TB_006              REQ_MINION_TARGET      0
28          CFM_621t21      REQ_MINIMUM_TOTAL_MINIONS      1
29          LOEA06_04h             REQ_TARGET_TO_PLAY      0
..                 ...                            ...    ...
746            GVG_011              REQ_MINION_TARGET      0
747            CFM_061             REQ_TARGET_TO_PLAY      0
748      AT_132_HUNTER                REQ_STEADY_SHOT      0
749      AT_132_HUNTER       REQ_MINION_OR_ENEMY_HERO      0
750            GVG_047              REQ_MINION_TARGET      0
751            GVG_047               REQ_ENEMY_TARGET      0
752            EX1_278             REQ_TARGET_TO_PLAY      0
753            CS2_012             REQ_TARGET_TO_PLAY      0
754            CS2_012               REQ_ENEMY_TARGET      0
755            EX1_578             REQ_TARGET_TO_PLAY      0
756            EX1_578              REQ_MINION_TARGET      0
757         KARA_00_10  REQ_SECRET_CAP_FOR_NON_SECRET      0
758         KARA_13_11      REQ_MINIMUM_ENEMY_MINIONS      2
759            GVG_120        REQ_TARGET_IF_AVAILABLE      0
760            GVG_120           REQ_TARGET_WITH_RACE     20
761         LOEA02_10a           REQ_NUM_MINION_SLOTS      1
762            CS1_130             REQ_TARGET_TO_PLAY      0
763             AT_025      REQ_MINIMUM_ENEMY_MINIONS      2
764            CS2_061             REQ_TARGET_TO_PLAY      0
765             AT_084            REQ_FRIENDLY_TARGET      0
766             AT_084        REQ_TARGET_IF_AVAILABLE      0
767             AT_084              REQ_MINION_TARGET      0
768           BRMA16_3            REQ_WEAPON_EQUIPPED      0
769           BRMA16_3             REQ_TARGET_TO_PLAY      0
770           BRMA16_3              REQ_MINION_TARGET      0
771            EX1_624             REQ_TARGET_TO_PLAY      0
772             AT_044             REQ_TARGET_TO_PLAY      0
773             AT_044              REQ_MINION_TARGET      0
774      AT_132_SHAMAN           REQ_NUM_MINION_SLOTS      1
775            CS2_031             REQ_TARGET_TO_PLAY      0

[776 rows x 3 columns]

In [25]:
cards_df.set_index("card_id", inplace=True)

In [26]:
dust_df.set_index('card_id', inplace=True)

In [28]:
entourages_df.set_index("card_id", inplace=True)

In [29]:
mechanics_df.set_index("card_id", inplace=True)

In [30]:
play_requirements_df.set_index('card_id', inplace=True)

Great.  Now, let's set the correct column, `card_id`, as the index column for each of these tables, and then display each to ensure that everything is as expected.  

For each of the dataframes we created in the cell above, call the `.set_index()` method and pass in `card_id`.  Also set `inplace=True`.  Then, display the head of each respective DataFrame to ensure everything worked.  

 - NOTE Since we are performing this operation in place, running any cell a second time will result in pandas throwing an error.  If you need to run something a second time, restart the kernel using the jupyter notebook menu at the top of the page.  

### Executing Joins

Now that we have the tables loaded correctly, we're going to execute some joins. There are 4 different kinds of joins, which can best be visualized with venn diagrams:

<img src='joins.png'>

In these diagrams, each circle represents a DataFrame or SQL Table.  The left table is the table we are working with, and the right table is the table we want to join to the table we are working with.  We'll start by executing the most common type of join, an **_Inner Join_**.

In the cell below, join `cards_df` with `mechanics_df` using the built-in `.join()` method on the `cards_df` object. 

Pass in the following parameters:
* the table we want to join with, `mechanics_df`
* The `how` parameter set to the type of join we want, `'inner'`

In [35]:
cards_with_mechanics_df = cards_df.join(mechanics_df, how ="inner")
cards_with_mechanics_df

player_class         type                        name  \
card_id                                                                      
AT_002                       MAGE        SPELL                      Effigy   
AT_005t                   NEUTRAL       MINION                        Boar   
AT_006                       MAGE       MINION            Dalaran Aspirant   
AT_007                       MAGE       MINION                Spellslinger   
AT_009                       MAGE       MINION                      Rhonin   
AT_010                     HUNTER       MINION                Ram Wrangler   
AT_011e                   NEUTRAL  ENCHANTMENT            Light's Blessing   
AT_012                     PRIEST       MINION            Spawn of Shadows   
AT_017                    NEUTRAL       MINION           Twilight Guardian   
AT_018                     PRIEST       MINION         Confessor Paletress   
AT_019                    WARLOCK       MINION                  Dreadsteed   
AT_022                    WARLOCK        SPELL            Fist of Jaraxxus   
AT_023                    WARLOCK       MINION                Void Crusher   
AT_028                      ROGUE       MINION             Shado-Pan Rider   
AT_030                      ROGUE       MINION           Undercity Valiant   
AT_032                      ROGUE       MINION                Shady Dealer   
AT_035t                     ROGUE        SPELL                     Ambush!   
AT_036                      ROGUE       MINION                   Anub'arak   
AT_037                      DRUID        SPELL                Living Roots   
AT_038                      DRUID       MINION          Darnassus Aspirant   
AT_038                      DRUID       MINION          Darnassus Aspirant   
AT_039                      DRUID       MINION            Savage Combatant   
AT_039e                     DRUID  ENCHANTMENT                      Savage   
AT_040                      DRUID       MINION                  Wildwalker   
AT_042                      DRUID       MINION          Druid of the Saber   
AT_042t                     DRUID       MINION             Sabertooth Lion   
AT_042t2                    DRUID       MINION          Sabertooth Panther   
AT_045                      DRUID       MINION                      Aviana   
AT_046                     SHAMAN       MINION             Tuskarr Totemic   
AT_047                     SHAMAN       MINION         Draenei Totemcarver   
...                           ...          ...                         ...   
TB_SPT_DPromoSecret3      WARRIOR        SPELL            Visions of Valor   
TB_SPT_DPromoSecret4      WARRIOR        SPELL             Visions of Fate   
TB_SPT_DPromoSecret5      WARRIOR        SPELL       Visions of the Amazon   
TB_SPT_DPromoSecret6      WARRIOR        SPELL     Visions of the Sorcerer   
TB_SPT_DPromoSecret7      WARRIOR        SPELL  Visions of the Necromancer   
TB_SPT_DPromoSecret9      WARRIOR        SPELL        Visions of Knowledge   
TB_SPT_DPromo_Hero2       WARRIOR         HERO                The Cow King   
TB_SPT_DpromoPortal       NEUTRAL       MINION            Enigmatic Portal   
TB_SPT_DpromoPortal       NEUTRAL       MINION            Enigmatic Portal   
TB_SPT_Minion1            NEUTRAL       MINION                  Shieldsman   
TB_SPT_Minion2            NEUTRAL       MINION             Battle Standard   
TB_SPT_Minion2            NEUTRAL       MINION             Battle Standard   
TU4c_003                  NEUTRAL       MINION                      Barrel   
TU4c_008e                 NEUTRAL  ENCHANTMENT              Might of Mukla   
TU4f_007                  NEUTRAL       MINION                Crazy Monkey   
XXX_001                   NEUTRAL        SPELL                    Damage 1   
XXX_002                   NEUTRAL        SPELL                    Damage 5   
XXX_008                   NEUTRAL        SPELL                      Freeze   
XXX_020                   NEUTRAL        SP

In [37]:
len(cards_with_mechanics_df)

1079

Examine the output from the cell above and compare it to the original output of both the `cards_df` and `mechanics_df` DataFrame heads we displayed earlier.  Notice how it this now combines the columns from both?

**_Question_**

If you inspect the original `cards_df` DataFrame, you'll notice that it contains  2,819 records.  The result of our inner join, `cards_with_mechanics_df`, conntains only 1079 rows.  Why?

Write your answer below this line:
________________________________________________________________________________

- Answer - the above inner join query combines all columns in which the the value of the card_id column was present in both the cards_df table/dataframe as well as the mechanics_df table/dataframe

### Other Types of Joins

By default, the `.join()` method performs a left join if no parameter is passed in for `how=`.  In the cell below, perform a **_Left Join_** of `cards_with_mechanics_df` and `play_requirements_df`, with `cards_with_mechanics_df` as the left table.  

Then, display `left_join_df` to inspect our results. 

In [6]:
left_join_df = cards_with_mechanics_df.join("play_requirements_df")
left_join_df

NameError: name 'cards_with_mechanics_df' is not defined

- Note that the results of this sort of join are dependent upon the position of each table--if we were to make `cards_with_mechanics_df` the right table and `play_requirements_df` the left table and then perform a **_Right Join_**, our results would be the same. 

**_Question:_**

Describe what was included from each table in this join.

Write your answer below this line:
________________________________________________________________________________
- Lee's answer - this joined all rows in the left_join_df with the rows in "play_requirements_df" where the primary_ids were found in left_join_df 


#### Outer Joins

In the cell below, perform an outer join between `cards_df` and `dust_df`. Since these tables contain columns with the same name, we'll need to specify a suffix for at least one of them, so that the column can be renamed to avoid a naming collision. 

During this join, set the `rsuffix` parameter to `_dust`

In [ ]:
outer_join_df = None
outer_join_df

Inspect the output above.  Note that the naming collision has been avoided by renaming the `cost` column from the right table to `cost_dust`.  

## Summary

In this lab, we learned how to:

* Concatenate multiple DataFrames together into a single DataFrame
* Use `sqlalchemy` to connect to a sqlite database and read in individual tables as DataFrames
* Understand and execute the various types of joins (inner, outer, left, and right joins)